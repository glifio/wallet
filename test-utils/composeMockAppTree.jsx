/* eslint-disable react/prop-types */
import React, { useReducer } from 'react'
import cloneDeep from 'lodash.clonedeep'

import WalletProvider from '@openworklabs/filecoin-wallet-provider'
import { FilecoinNumber } from '@openworklabs/filecoin-number'
import { Provider } from 'react-redux'

import { NetworkCheck } from '../lib/check-network'
import BalancePoller from '../lib/update-balance'
import { ConverterContext } from '../lib/Converter'
import { theme, ThemeProvider } from '../components/Shared'
import { initializeStore } from './index'
import { initialState } from '../store/states'
import walletProviderReducer, {
  initialState as walletProviderInitialState,
  setWalletType,
  setError,
  resetLedgerState,
  resetState
} from '../WalletProvider/state'
import { WalletProviderContext } from '../WalletProvider'
import createPath from '../utils/createPath'
import { IMPORT_MNEMONIC } from '../constants'

const useRouter = jest.spyOn(require('next/router'), 'useRouter')

const mockRouter = jest.fn(() => {})
useRouter.mockImplementation(() => ({
  query: 'network=t',
  push: mockRouter,
  replace: mockRouter
}))

jest.mock('@openworklabs/filecoin-wallet-provider')
const mockGetAccounts = jest
  .fn()
  .mockImplementation(() =>
    Promise.resolve(['t1mbk7q6gm4rjlndfqw6f2vkfgqotres3fgicb2uq'])
  )

const mockGetBalance = jest
  .fn()
  .mockImplementation(() => Promise.resolve(new FilecoinNumber('1', 'fil')))

WalletProvider.mockImplementation(() => {
  return {
    wallet: {
      getAccounts: mockGetAccounts,
      sign: jest.fn().mockImplementation(() => 'xxxyyyyzzzz')
    },
    getBalance: mockGetBalance,
    getNonce: jest.fn().mockImplementation(() => 0),
    estimateGas: jest
      .fn()
      .mockImplementation(() => new FilecoinNumber('122', 'attofil')),
    sendMessage: jest.fn().mockImplementation(() => 'QmZCid!')
  }
})

const mockWalletProviderInstance = new WalletProvider()

const presets = {
  preOnboard: cloneDeep(initialState),
  postOnboard: cloneDeep({
    ...initialState,
    wallets: [
      {
        address: 't1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi',
        balance: new FilecoinNumber('1', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0
  })
}

const composeWalletProviderState = (initialWalletProviderState, preset) => {
  switch (preset) {
    case 'postOnboard': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletType: IMPORT_MNEMONIC,
        walletProvider: mockWalletProviderInstance
      })
    }
    default:
      return initialWalletProviderState
  }
}

/**
 * VALID OPTIONS:
 * mockConverterInstance (an object that stubs the converter functionality)
 * walletProviderReducer (a reducer to stub the state handling of the wallet provider)
 * walletProviderInitialState(a reducer to stub the state of the walletProvider context)
 * pathname
 * query (usually { network: 'f' } or { network: 't' })
 */

export default (statePreset = 'preOnboard', options = {}) => {
  const state = options.state || presets[statePreset]
  const store = initializeStore(state)

  const mockConverterInstance = options.mockConverterInstance || {
    cacheConversionRate: jest.fn(),
    toFIL: jest.fn().mockImplementation(amount => {
      return new FilecoinNumber(amount, 'fil').dividedBy(5)
    }),
    fromFIL: jest.fn().mockImplementation(amount => {
      return new FilecoinNumber(amount, 'fil').multipliedBy(5)
    })
  }

  const pathname = options.pathname || '/wallet'
  const query = options.query || { network: 't' }

  const Tree = ({ children }) => {
    const [initialWalletProviderState, walletProviderDispatch] = useReducer(
      options.reducer || walletProviderReducer,
      options.walletProviderInitialState || walletProviderInitialState
    )

    const walletProviderState = composeWalletProviderState(
      initialWalletProviderState,
      statePreset
    )
    return (
      <Provider store={store}>
        <ConverterContext.Provider
          value={{
            converter: mockConverterInstance,
            converterError: options.converterError || null
          }}
        >
          <WalletProviderContext.Provider
            value={{
              state: walletProviderState,
              dispatch: walletProviderDispatch,
              fetchDefaultWallet: jest.fn().mockImplementation(),
              setWalletError: jest
                .fn()
                .mockImplementation(errorMessage =>
                  walletProviderDispatch(setError(errorMessage))
                ),
              setWalletType: jest
                .fn()
                .mockImplementation(walletType =>
                  walletProviderDispatch(setWalletType(walletType))
                ),
              setLedgerProvider: jest.fn(),
              connectLedger: jest.fn(),
              resetLedgerState: jest.fn().mockImplementation(() => {
                walletProviderDispatch(resetLedgerState())
              }),
              resetState: jest.fn().mockImplementation(() => {
                walletProviderDispatch(resetState())
              })
            }}
          >
            <NetworkCheck
              networkFromRdx={store.getState().network}
              pathname={pathname}
              query={query}
              switchNetwork={jest.fn()}
            />
            <BalancePoller />
            <ThemeProvider theme={theme}>{children}</ThemeProvider>
          </WalletProviderContext.Provider>
        </ConverterContext.Provider>
      </Provider>
    )
  }

  return { Tree, store, walletProvider: mockWalletProviderInstance }
}
