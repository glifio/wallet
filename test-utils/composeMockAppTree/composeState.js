import cloneDeep from 'lodash.clonedeep'
import { FilecoinNumber } from '@openworklabs/filecoin-number'

import createPath from '../../utils/createPath'
import { IMPORT_MNEMONIC } from '../../constants'
import { initialState } from '../../store/states'
import { mockWalletProviderInstance } from '../mocks/mock-wallet-provider'

export const presets = {
  preOnboard: cloneDeep(initialState),
  postOnboard: cloneDeep({
    ...initialState,
    wallets: [
      {
        address: 't1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi',
        balance: new FilecoinNumber('1', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0
  }),
  postOnboardWithError: cloneDeep({
    ...initialState,
    error: 'error for testing',
    wallets: [
      {
        address: 't1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi',
        balance: new FilecoinNumber('1', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0
  })
}

export const composeWalletProviderState = (
  initialWalletProviderState,
  preset
) => {
  switch (preset) {
    case 'postOnboard': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletType: IMPORT_MNEMONIC,
        walletProvider: mockWalletProviderInstance
      })
    }
    case 'postOnboardWithError': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletType: IMPORT_MNEMONIC,
        walletProvider: mockWalletProviderInstance
      })
    }
    default:
      return initialWalletProviderState
  }
}
