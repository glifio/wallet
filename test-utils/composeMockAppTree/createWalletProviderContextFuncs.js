import { FilecoinNumber } from '@glif/filecoin-number'

import {
  setWalletType,
  setError,
  resetLedgerState,
  resetState
} from '../../WalletProvider/state'

export default walletProviderDispatch => ({
  fetchDefaultWallet: jest.fn().mockImplementation(() => ({
    balance: new FilecoinNumber('1', 'fil'),
    address: 't1mbk7q6gm4rjlndfqw6f2vkfgqotres3fgicb2uq',
    path: "m/44'/1'/0/0/0"
  })),
  setWalletError: jest
    .fn()
    .mockImplementation(errorMessage =>
      walletProviderDispatch(setError(errorMessage))
    ),
  setWalletType: jest
    .fn()
    .mockImplementation(walletType =>
      walletProviderDispatch(setWalletType(walletType))
    ),
  setLedgerProvider: jest.fn(),
  connectLedger: jest.fn(),
  resetLedgerState: jest.fn().mockImplementation(() => {
    walletProviderDispatch(resetLedgerState())
  }),
  resetState: jest.fn().mockImplementation(() => {
    walletProviderDispatch(resetState())
  })
})
