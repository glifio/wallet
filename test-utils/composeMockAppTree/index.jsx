/* eslint-disable react/prop-types */
import React, { useReducer } from 'react'
import { Provider } from 'react-redux'
import { Converter } from '@openworklabs/filecoin-number'

import { NetworkCheck } from '../../lib/check-network'
import BalancePoller from '../../lib/update-balance'
import { ConverterContext } from '../../lib/Converter'
import { theme, ThemeProvider } from '../../components/Shared'
import { initializeStore } from '..'
import walletProviderReducer, {
  initialState as walletProviderInitialState
} from '../../WalletProvider/state'
import { WasmContext } from '../../lib/WasmLoader'
import { WalletProviderContext } from '../../WalletProvider'
import { mockWalletProviderInstance } from '../mocks/mock-wallet-provider'
import createMockWalletProviderContextFuncs from './createWalletProviderContextFuncs.js'
import mockWalletSubproviders from '../mocks/mock-wallet-subproviders'
import { presets, composeWalletProviderState } from './composeState'

/**
 * VALID OPTIONS:
 * mockConverterInstance (an object that stubs the converter functionality)
 * walletProviderReducer (a reducer to stub the state handling of the wallet provider)
 * walletProviderInitialState(a reducer to stub the state of the walletProvider context)
 * walletProviderDispatch(a dispatcher to mock/read calls to the walletProvider's state)
 * pathname
 * query (usually { network: 'f' } or { network: 't' })
 */

export default (statePreset = 'preOnboard', options = {}) => {
  const state = options.state || presets[statePreset]
  const store = initializeStore(state)

  const pathname = options.pathname || '/wallet'
  const query = options.query || { network: 't' }

  const Tree = ({ children }) => {
    const [initialWalletProviderState, walletProviderDispatch] = useReducer(
      options.reducer || walletProviderReducer,
      options.walletProviderInitialState || walletProviderInitialState
    )

    const walletProviderState = composeWalletProviderState(
      initialWalletProviderState,
      statePreset
    )

    const mockWalletProviderContextFuncs = createMockWalletProviderContextFuncs(
      options.walletProviderDispatch || walletProviderDispatch
    )

    return (
      <Provider store={store}>
        <WasmContext.Provider
          value={{
            generateMnemonic: jest
              .fn()
              .mockImplementation(
                () =>
                  'slender spread awkward chicken noise useful thank dentist tip bronze ritual explain version spot collect whisper glow peanut bus local country album punch frown'
              )
          }}
        >
          <ConverterContext.Provider
            value={{
              converter: new Converter(),
              converterError: options.converterError || null
            }}
          >
            <WalletProviderContext.Provider
              value={{
                state: walletProviderState,
                dispatch:
                  options.walletProviderDispatch || walletProviderDispatch,
                walletSubproviders: {
                  ...mockWalletSubproviders
                },
                ...mockWalletProviderContextFuncs
              }}
            >
              <NetworkCheck
                networkFromRdx={store.getState().network}
                pathname={pathname}
                query={query}
                switchNetwork={jest.fn()}
              />
              <BalancePoller />
              <ThemeProvider theme={theme}>{children}</ThemeProvider>
            </WalletProviderContext.Provider>
          </ConverterContext.Provider>
        </WasmContext.Provider>
      </Provider>
    )
  }

  return {
    Tree,
    store,
    walletProvider: mockWalletProviderInstance
  }
}
