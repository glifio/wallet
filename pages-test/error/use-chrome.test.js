import { cleanup, render, screen, act, fireEvent } from '@testing-library/react'

import UseChrome from '../../pages/error/use-chrome.jsx'
import composeMockAppTree from '../../test-utils/composeMockAppTree'

jest.mock('@openworklabs/filecoin-wallet-provider')

describe('UseChrome', () => {
  afterEach(cleanup)
  test('it renders the error page', () => {
    const { Tree } = composeMockAppTree('postOnboard')
    const { container } = render(
      <Tree>
        <UseChrome />
      </Tree>
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  test('it renders the message about getting Chrome', () => {
    const { Tree } = composeMockAppTree('postOnboard')

    render(
      <Tree>
        <UseChrome />
      </Tree>
    )

    expect(
      screen.getByText(
        'Please install Google Chrome to continue using your Ledger device, or choose an alternative setup option'
      )
    ).toBeInTheDocument()
  })

  test('it renders the home page after clicking back', async () => {
    const { Tree } = composeMockAppTree('preOnboard')
    const useRouter = jest.spyOn(require('next/router'), 'useRouter')
    const mockRouterReplace = jest.fn(() => {})
    useRouter.mockImplementationOnce(() => ({
      query: 'network=t',
      replace: mockRouterReplace
    }))

    let res
    await act(async () => {
      res = render(
        <Tree>
          <UseChrome />
        </Tree>
      )
      fireEvent.click(screen.getByText('Back'))
    })

    expect(mockRouterReplace).toHaveBeenCalled()
    expect(mockRouterReplace).toHaveBeenCalledWith('/onboard')
  })
})
