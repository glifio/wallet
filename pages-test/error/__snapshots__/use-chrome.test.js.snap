// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UseChrome it renders the error page 1`] = `
.c13 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
}

.c13 svg path {
  stroke: var(--black);
}

.c13:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c13:hover svg path {
  stroke: var(--purple-medium);
}

.c13:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c13:active svg path {
  stroke: var(--white) !important;
}

.c4 {
  width: auto;
  height: 35px;
}

.c15 {
  width: auto;
  height: 35px;
  margin-bottom: 2em;
}

.c10 {
  padding: 1.5em;
  border-radius: 8px;
  overflow: hidden;
  text-align: center;
  word-break: break-word;
}

.c10 > *:first-child {
  margin-top: 0;
}

.c10 > *:last-child {
  margin-bottom: 0;
}

.c10 > svg:first-child {
  display: block;
  margin: 0 auto 1em;
}

.c10 > h2:first-child {
  margin: 0;
}

.c10 header:first-child {
  margin: -1.5em -1.5em 1.5em -1.5em;
  padding: 1.5em;
  background-color: var(--purple-medium);
  color: var(--white);
}

.c11 {
  background-color: var(--red-light);
  color: var(--red-dark);
}

.c11 hr {
  border-color: var(--red-dark);
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: var(--space-m);
  margin: 1.5em 0;
}

.c7 {
  position: relative;
  padding: 3rem 1.5rem;
}

.c7:not(:last-child) {
  padding-bottom: 0;
}

.c7 > *:first-child {
  margin-top: 0;
}

.c7 > *:last-child {
  margin-bottom: 0;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: var(--space-m);
  width: 100%;
  margin: 0 auto;
  max-width: 40em;
}

.c9 form {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: var(--space-m);
}

.c2 {
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  z-index: 100;
  padding: var(--space-m) 0;
  margin: var(--space-nm) 0;
  background-color: var(--white-broken);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-column-gap: var(--space-l);
  column-gap: var(--space-l);
  row-gap: var(--space-m);
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  gap: var(--space-l);
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c5 {
  gap: 0;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c17 {
  list-style-type: none;
  margin: 0;
  padding-left: 0;
}

.c18 {
  list-style-type: none;
  margin: 0;
  padding-left: 0;
}

.c14 {
  background: #272727;
  border-radius: 8px;
  padding: 30px;
  color: white;
  position: relative;
}

.c20 {
  text-transform: uppercase;
  display: block;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  max-width: 1920px;
  margin: 0 auto;
  padding: var(--space-m);
  gap: var(--space-m);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: var(--space-m);
}

.c1 > * {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c1 > *:last-child {
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
}

.c6 {
  min-height: 0;
}

@media (min-width:480px) {
  .c4 {
    height: 45px;
  }
}

@media (min-width:868px) {
  .c4 {
    height: 55px;
  }
}

@media (min-width:480px) {
  .c15 {
    height: 45px;
  }
}

@media (min-width:868px) {
  .c15 {
    height: 55px;
    position: absolute;
    left: 30px;
    top: 30px;
  }
}

@media (min-width:868px) {
  .c7 {
    padding: 3rem;
  }
}

@media (min-width:868px) {
  .c16 {
    max-width: 670px;
    display: grid;
    grid-template-columns: repeat(3,1fr);
    gap: var(--space-m);
    margin: 0 auto;
  }
}

@media (max-width:868px) {
  .c19 {
    margin-top: 2em;
  }
}

@media (min-width:868px) {
  .c20 {
    margin-top: 1em;
  }
}

@media (max-width:868px) {
  .c20 {
    margin-top: 5em;
  }
}

@media (max-width:868px) {

}

@media (min-width:868px) {

}

@media (min-width:868px) {
  .c1 {
    min-height: calc(100vh - 2 * var(--space-m));
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <header
      class="c2"
    >
      <nav
        class="c3"
      >
        <a
          href="/"
        >
          <svg
            class="c4"
            fill="none"
            height="960"
            viewBox="0 0 640 960"
            width="640"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 0H630C635.523 0 640 4.47715 640 10V70C640 75.5229 635.523 80 630 80H385C371.193 80 360 91.1929 360 105V290C360 295.523 355.523 300 350 300H290C284.477 300 280 295.523 280 290V105C280 91.1929 268.807 80 255 80H10C4.47718 80 3.05176e-05 75.5229 3.05176e-05 70V10C3.05176e-05 4.47715 4.47718 0 10 0ZM420 710C420 726.569 406.569 740 390 740C373.431 740 360 726.569 360 710V370C360 364.477 355.523 360 350 360H290C284.477 360 280 364.477 280 370V710C280 726.569 266.569 740 250 740C233.431 740 220 726.569 220 710L220 150C220 144.477 215.523 140 210 140L150 140C144.477 140 140 144.477 140 150L140 749.976L140 750V810C140 815.523 144.477 820 150 820H210H290H350H430H490C495.523 820 500 815.523 500 810V750V150C500 144.477 495.523 140 490 140L430 140C424.477 140 420 144.477 420 150V710ZM370 880H535C548.807 880 560 868.807 560 855V150C560 144.477 564.477 140 570 140H630C635.523 140 640 144.477 640 150V890V950C640 955.523 635.523 960 630 960H570H370C364.477 960 360 955.523 360 950V890C360 884.477 364.477 880 370 880ZM2.89871e-05 889.976L0 890V950C0 952.071 0.629602 953.995 1.70785 955.591C3.50492 958.251 6.54823 960 10 960H70H270C275.523 960 280 955.523 280 950V890C280 884.477 275.523 880 270 880H105C91.1929 880 80 868.807 80 855L80 150C80 144.477 75.5229 140 70 140H10C4.47717 140 1.26938e-05 144.477 1.28145e-05 150L2.89871e-05 889.976Z"
              fill="black"
              fill-rule="evenodd"
            />
          </svg>
        </a>
      </nav>
      <nav
        class="c3 c5"
      />
    </header>
    <div
      class="c1 c6"
    >
      <div
        class="c7 c8"
      >
        <div
          class="c9"
        >
          <div
            class="c10 c11"
          >
            <svg
              fill="none"
              height="20px"
              viewBox="0 0 22 20"
              width="22px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11 7.74982V9.99986M11 14.4999H11.0112M3.24282 19H18.7572C20.4815 19 21.5586 17.1246 20.6965 15.6249L12.9393 2.12474C12.0771 0.625088 9.92286 0.625088 9.0607 2.12474L1.30352 15.6249C0.441363 17.1246 1.5185 19 3.24282 19Z"
                stroke="var(--red-dark)"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <h2>
              Only Chrome has Ledger support
            </h2>
            <hr />
            <p>
              Please install Google Chrome to continue using your Ledger device, or choose an alternative setup option
            </p>
            <p>
              <a
                href="https://www.google.com/chrome"
                rel="noreferrer noopener"
                target="_blank"
              >
                Install Google Chrome.
              </a>
            </p>
            <hr />
            <p>
              Get help in 
              <a
                href="https://discord.gg/B9ju5Eu4Rq"
                rel="noreferrer noopener"
                target="_blank"
              >
                Discord
              </a>
               or hit us up on 
              <a
                href="https://twitter.com/glifio"
                rel="noreferrer noopener"
                target="_blank"
              >
                Twitter
              </a>
            </p>
          </div>
          <div
            class="c12"
          >
            <button
              class="c13"
            >
              Back
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="c14"
  >
    <svg
      class="c15"
      height="71px"
      version="1.1"
      viewBox="0 0 52 71"
      width="52px"
      xmlns="http://www.w3.org/2000/svg"
      xmlns:xlink="http://www.w3.org/1999/xlink"
    >
      <title>
        logo-glif
      </title>
      <g
        fill="none"
        fill-rule="evenodd"
        id="Page-1"
        stroke="none"
        stroke-width="1"
      >
        <g
          fill="#FFFFFF"
          fill-rule="nonzero"
          id="logo-glif"
          transform="translate(-0.000000, 0.000000)"
        >
          <path
            d="M10.4352,26.1836 C10.4352,26.2474 10.4481,26.3105 10.4727,26.3693 C10.4973,26.4281 10.5332,26.4815 10.5785,26.5264 C10.6237,26.5714 10.6773,26.6069 10.7363,26.6312 C10.7952,26.6554 10.8584,26.6675024 10.9222,26.6675024 L50.5292,26.6675024 C50.593,26.6675024 50.6561,26.6554 50.7151,26.6312 C50.7741,26.6069 50.8277,26.5714 50.8729,26.5264 C50.9181,26.4815 50.9541,26.4281 50.9786,26.3693 C51.0032,26.3105 51.0162,26.2474 51.0162,26.1836 L51.0162,10.919 C51.0162,10.8552 51.0032,10.7922 50.9786,10.7333 C50.9541,10.6745 50.9181,10.6211 50.8729,10.5762 C50.8277,10.5313 50.7741,10.4957 50.7151,10.4715 C50.6561,10.4473 50.593,10.4349 50.5292,10.4351 L21.3573,10.4351 C21.2935,10.4349 21.2304,10.4473 21.1714,10.4715 C21.1124,10.4957 21.0588,10.5313 21.0136,10.5762 C20.9683,10.6211 20.9324,10.6745 20.9078,10.7333 C20.8833,10.7922 20.8705,10.8552 20.8703,10.919 L20.8703,15.7485 C20.8705,15.8123 20.8833,15.8753 20.9078,15.9342 C20.9324,15.993 20.9683,16.0464 21.0136,16.0913 C21.0588,16.1362 21.1124,16.1718 21.1714,16.196 C21.2304,16.2202 21.2935,16.2326 21.3573,16.2324 L42.9,16.2324 C43.515,16.2324 44.1048,16.4767 44.5397,16.9116 C44.9746,17.3465 45.2189,17.9363 45.2189,18.5513 C45.2189,19.1663 44.9746,19.7562 44.5397,20.191 C44.1048,20.6259 43.515,20.8702 42.9,20.8702 L18.5514,20.8702 C17.9364,20.8702 17.3465,20.6259 16.9117,20.191 C16.4768,19.7562 16.2325,19.1663 16.2325,18.5513 L16.2325,8.1162 C16.2325,7.50119 16.4768,6.91136 16.9117,6.47648 C17.3465,6.0416 17.9364,5.79729 18.5514,5.79729 L50.5292,5.79729 C50.593,5.79749 50.6561,5.78513 50.7151,5.76093 C50.7741,5.73672 50.8277,5.70113 50.8729,5.6562 C50.9181,5.61127 50.9541,5.55787 50.9786,5.49905 C51.0032,5.44024 51.0162,5.37715 51.0162,5.31341 L51.0162,0.483881 C51.0156,0.355201 50.964,0.232012 50.8727,0.141311 C50.7814,0.0506101 50.6579,0 50.5292,0 L10.9222,0 C10.7935,0 10.67,0.0506101 10.5787,0.141311 C10.4874,0.232012 10.4358,0.355201 10.4352,0.483881 L10.4352,26.1836 Z"
            id="Path"
          />
          <path
            d="M50.5284,47.6879024 C50.6572,47.6883 50.7809,47.6376 50.8723,47.5468 C50.9638,47.4561 51.0155,47.3328 51.0161,47.204 L51.0161,42.3706 C51.0155,42.2418 50.9638,42.1185 50.8723,42.0278 C50.7809,41.937 50.6572,41.8863 50.5284,41.8866977 L0.487747,41.8866977 C0.358933,41.8863 0.235219,41.937 0.143772,42.0278 C0.0523238001,42.1185 0.000616192051,42.2418 5.25659516e-11,42.3706 L5.25659516e-11,47.2001 C5.25659516e-11,47.2642 0.0122749001,47.3278 0.0367141001,47.387 C0.0611532001,47.4463 0.0970723001,47.5001 0.142404,47.5455 C0.187736,47.5908 0.241586,47.6267 0.300854,47.6512 C0.360122,47.6756 0.423639,47.6881 0.487747,47.6879024 L50.5284,47.6879024 Z"
            id="Path"
          />
          <path
            d="M50.5284,37.2413054 C50.5921,37.2413054 50.6553,37.2293 50.7143,37.2052 C50.7733,37.1811 50.827,37.1456 50.8723,37.1008 C50.9176,37.0559 50.9536,37.0025 50.9783,36.9438 C51.0029,36.885 51.0161,36.8219 51.0161,36.7582 L51.0161,31.9279 C51.0155,31.7991 50.9638,31.6758 50.8723,31.585 C50.7809,31.4943 50.6572,31.4436 50.5284,31.444 L8.1162,31.444 C7.50119,31.444 6.91136,31.1997 6.47648,30.7648 C6.0416,30.3299 5.79729,29.7401 5.79729,29.1251 L5.79729,0.626417 C5.7977,0.497602 5.74697,0.373889 5.65625,0.282441 C5.56553,0.190993 5.44222,0.139286 5.31341,0.138672 L0.483882,0.138672 C0.355068,0.139286 0.231763,0.190993 0.14104,0.282441 C0.0503165001,0.373889 5.71134251e-11,0.497602 5.71134251e-11,0.626417 L5.71134251e-11,36.7582 C5.71134251e-11,36.8811 0.0473625001,36.9993 0.13218,37.0882 L0.143002,37.0998 C0.187479,37.1443 0.240271,37.1795 0.298363,37.2035 C0.356455,37.2275 0.418707,37.2398 0.481564,37.2397 L50.5284,37.2397 L50.5284,37.2413054 Z"
            id="Path"
          />
          <path
            d="M0,70.3858 C-0.000684023949,70.4499 0.0112877001,70.5136 0.0352565001,70.573 C0.0592253001,70.6325 0.0947178001,70.6866 0.139691,70.7323 C0.184664,70.778 0.23823,70.8143 0.297306,70.8393 C0.356381,70.8642 0.419799,70.8771 0.48391,70.8774 L5.31344,70.8774 C5.44225,70.8768 5.56556,70.8251 5.65628,70.7337 C5.747,70.6422 5.79773,70.5185 5.79732,70.3897 C5.79732,69.904 5.99024,69.4383 6.33365,69.0949 C6.67706,68.7515 7.14283,68.5585 7.62848,68.5585 L50.5284,68.5585 C50.6572,68.5589 50.7809,68.5082 50.8724,68.4175 C50.9638,68.3268 51.0155,68.2035 51.0161,68.0746 L51.0161,63.2413 C51.0155,63.1124 50.9638,62.9891 50.8724,62.8984 C50.7809,62.8077 50.6572,62.757 50.5284,62.7574 L8.11623,62.7574 C7.50121,62.7574 6.91139,62.5131 6.47651,62.0782 C6.04163,61.6433 5.79732,61.0535 5.79732,60.4385 C5.79732,59.8235 6.04163,59.2336 6.47651,58.7987 C6.91139,58.3639 7.50121,58.1196 8.11623,58.1196 L50.5284,58.1196 C50.6572,58.12 50.7809,58.0692 50.8724,57.9785 C50.9638,57.8878 51.0155,57.7645 51.0161,57.6357 L51.0161,52.8061 C51.0155,52.6773 50.9638,52.554 50.8724,52.4633 C50.7809,52.3726 50.6572,52.3219 50.5284,52.3222977 L0.487775,52.3222977 C0.35896,52.3219 0.235247,52.3726 0.143799,52.4633 C0.0523515001,52.554 0.000643898051,52.6773 0,52.8061 L0,70.3858 Z"
            id="Path"
          />
        </g>
      </g>
    </svg>
    <div
      class="c16"
    >
      <ul
        class="c17"
      >
        <li
          class="c18"
        >
          <a
            href="https://discord.gg/B9ju5Eu4Rq"
            rel="noreferrer noopener"
            target="_blank"
          >
            Discord
          </a>
        </li>
        <li
          class="c18"
        >
          <a
            href="https://blog.glif.io"
            rel="noreferrer noopener"
            target="_blank"
          >
            Blog
          </a>
        </li>
        <li
          class="c18"
        >
          <a
            href="https://github.com/glifio"
            rel="noreferrer noopener"
            target="_blank"
          >
            Code
          </a>
        </li>
        <li
          class="c18"
        >
          <a
            href="https://lotus.filecoin.io/docs/developers/hosted-lotus/"
            rel="noreferrer noopener"
            target="_blank"
          >
            Nodes
          </a>
        </li>
        <li
          class="c18"
        >
          <a
            href="mailto:squad@infinitescroll.org"
            rel="noreferrer noopener"
            target="_self"
          >
            Contact
          </a>
        </li>
      </ul>
      <ul
        class="c17"
      >
        <li
          class="c18"
        >
          <a
            href="https://wallet.glif.io"
            rel="noreferrer noopener"
            target="_blank"
          >
            Wallet
          </a>
        </li>
        <li
          class="c18"
        >
          <a
            href="https://safe.glif.io"
            rel="noreferrer noopener"
            target="_blank"
          >
            Safe
          </a>
        </li>
        <li
          class="c18"
        >
          <a
            href="https://explorer.glif.io"
            rel="noreferrer noopener"
            target="_blank"
          >
            Explorer
          </a>
        </li>
        <li
          class="c18"
        >
          <a
            href="https://verify.glif.io"
            rel="noreferrer noopener"
            target="_blank"
          >
            Verifier
          </a>
        </li>
      </ul>
      <div
        class="c19"
      >
        GLIF
        <br />
        Freeport, Grand Bahamas
      </div>
      <div
        class="c20"
      >
        2022 Glif for life
      </div>
    </div>
  </footer>
</div>
`;
